client/asset/eth/eth.go

```diff
@@ -2353,24 +2440,157 @@ func (eth *ETHWallet) checkForNewBlocks(reportErr func(error)) {
 	}
 
 	eth.tipMtx.Lock()
-	defer eth.tipMtx.Unlock()
-
 	prevTip := eth.currentTip
 	eth.currentTip = bestHdr
+	eth.tipMtx.Unlock()
 
 	eth.log.Debugf("tip change: %s (%s) => %s (%s)", prevTip.Number,
 		currentTipHash, bestHdr.Number, bestHash)
 
+	walletList := eth.walletList()
+
 	go func() {
-		for _, w := range eth.walletList() {
+		for _, w := range walletList {
 			w.tipChange(nil)
 		}
 	}()
 	go func() {
-		for _, w := range eth.walletList() {
+		for _, w := range walletList {
 			w.checkFindRedemptions()
 		}
 	}()
+	go func() {
+		for _, w := range walletList {
+			w.checkConfirmRedemptions()
+		}
+	}()
+}
+
+// swapIsComplete returns true if the swap is either in the redeemed or
+// refunded state. This means that no further actions can be taken on
+// this swap.
+func (w *assetWallet) swapIsComplete(secretHash [32]byte, contractVersion uint32) (bool, error) {
+	swap, err := w.swap(w.ctx, secretHash, contractVersion)
+	if err != nil {
+		return false, err
+	}
+	return swap.State == dexeth.SSRedeemed || swap.State == dexeth.SSRefunded, nil
+}
+
+// confirmRedemption checks one of the queued redemption requests.
+func (w *assetWallet) confirmRedemption(txHash common.Hash, req *confirmRedemptionRequest, currentTip int64) {
+	const (
+		txConfsNeededToConfirm               = 10
+		blocksToWaitBeforeCoinNotFound       = 10
+		blocksToWaitBeforeCheckingIfReplaced = 10
+	)
+
+	var blocksSinceRequestStart int64
+	if currentTip > req.startHeight {
+		blocksSinceRequestStart = currentTip - req.startHeight
+	}
+
+	// We do not check that the tx is actually a redemption tx that redeems the
+	// expected secret hash because we are confirming the user's own redemption
+	// and not the counterparty's.
+	tx, txBlock, err := w.node.getTransaction(w.ctx, txHash)
+	if errors.Is(err, asset.CoinNotFoundError) {
+		w.log.Errorf("confirmRedemption for tx %s: could not find coin", txHash)
+		// Wait a few blocks before returning that the coin cannot be found.
+		if blocksSinceRequestStart < blocksToWaitBeforeCoinNotFound {
+			return
+		}
+		complete, err := w.swapIsComplete(req.secretHash, req.contractVersion)
+		if err != nil {
+			req.res <- err
+			return
+		}
+		if complete {
+			// the tx was replaced, and we don't know the tx, but it doesn't
+			// matter now.
+			req.res <- nil
```

Comment:
It **does** matter for error messaging. Looking for a redemption and finding a a refund is a notable event. Can we resolve this and the `TODO: handle this differently if the swap is refunded` issue now with a constant error and an error message and notification from core.

Reply:
If it is refunded it will return an error notification now. This message now is talking about if the swap is redeemed by a transaction the client is not aware of.